### ISSUE TITLE
LISA: add lisa_train_embeddings so the memory half of the claim is reachable (embeddings + head + final norm are 70.7% of what LISA trains at 8B)

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#377]
## What

LISA's documented claim is "full-FT quality at LoRA-like memory". The H100 session measured it and **the quality half holds while the memory half does not** — that correction shipped in v0.73.0's docs. This issue is the code change that would actually deliver the memory half.

## The measurement

Llama-3.1-8B, 3 interleaved repeats, engagement verified three ways rather than trusted from the flag (rotating layer set on cadence, trainable-parameter count matching the arithmetic to the unit, `lisa_num_layers = all` reproducing the full-FT arm to three decimals) — `benchmarks/gate-h100-validation.md` L3411–3442:

| arm | peak VRAM | held-out loss |
|---|---|---|
| full FT | **OOM at 73.94 GB** (all 3 repeats, even at batch 1) | — |
| LISA (2/20) | 52.14 GB | **1.294** |
| LoRA r=16 | 34.56 GB | 1.275 |

At 3B, each arm at its own better LR: full-FT 57.60 GB / 1.2905, LISA 19.37 GB / 1.2463, LoRA 15.93 GB / 1.2420.

**LISA beats full fine-tuning at both learning rates.** But it costs **1.22x LoRA's VRAM at 3B and 1.51x at 8B, and the gap widens with size.**

## The cause, which is what makes this fixable

Embeddings + LM head + final norm are trainable **every interval**, and at 8B they are:

```
1,486,901,248 trainable  =  1,050,677,248 always-on  +  2 x 218,112,000 sampled layers
```

**70.7% of everything LISA trains is the always-on set.** `lisa_num_layers` controls only ~30% of the cost, which is why turning it down does not buy what a user expects.

## Proposal

`training.lisa_train_embeddings: bool = true` — `false` freezes embeddings / LM head / final norm and leaves only the sampled decoder layers trainable.

This is a real trade, not a free win: the always-on set is presumably load-bearing for LISA's quality result, and turning it off may move the loss. That is exactly why it should be a knob with a measured answer rather than a silent default change.

## Rider to keep in the docs

The shipped defaults are the best points measured — raising `lisa_num_layers` degrades memory, speed **and** quality monotonically. LISA's real win as measured is: 8B trains on one 80 GB card where full FT needs ~120 GB.

## Acceptance criteria

- [ ] `lisa_train_embeddings: false` freezes the three groups, and the **trainable-parameter count matches the arithmetic to the unit** — the count is the assertion, not "the flag was accepted"
- [ ] Peak VRAM measured both ways on the same model, published
- [ ] Held-out loss measured both ways — if quality collapses, that is the result and it should be documented, not hidden
- [ ] Footgun rejection: set while `lisa_enabled: false`

**Comment by Amix29** [comment]
I'd like to take this issue. I can implement the scoped `training.lisa_train_embeddings` control and validate both the exact trainable-parameter accounting and a live same-model memory/loss comparison on Apple Silicon. I will keep the default behavior unchanged, add the disabled-LISA footgun rejection, and wait for confirmation here before opening the PR.

**Comment by MakazhanAlpamys** [comment]
Merged #576, contributed by @Srinivasan8888. It delivers the configurable LISA freeze control with parameter-count regression coverage and documented preflight behavior. The VRAM and held-out-loss measurements remain open acceptance work for this issue.

**Comment by Amix29** [comment]
@MakazhanAlpamys I completed the remaining live Apple Silicon comparison against current
`main` (`95b0fb6`) after #576. This is a clean rerun: the previously concurrent large Qwen
workload was no longer resident, and swap did not grow during either arm.

### Paired result

M4 Max, 128 GiB unified memory; `Qwen/Qwen2.5-7B-Instruct` at revision
`a09a35458c702b33eeacc393d103063234e8bc28`; 40 LISA steps; 2 sampled decoder layers;
rotation every 20 steps; batch size 1; sequence length 256; seed/data seed 377. Both arms
used the same 128-example dataset (SHA-256
`c46f8bad9aa2caad9b6b238445e2ec85e3cf5f48a7a509d29bd27a0a7456cf09`) and the exact
same sampled-layer schedule: `[12, 27]`, `[13, 24]`, then `[1, 8]`.

| Metric | `lisa_train_embeddings: true` | `false` | Change with `false` |
| --- | ---: | ---: | ---: |
| Trainable parameters | 1,556,113,920 | 466,115,584 | -70.05% |
| Peak MPS driver allocation | 113.50 GiB | 40.46 GiB | -73.05 GiB (-64.36%) |
| Peak MPS current allocation | 47.82 GiB | 34.21 GiB | -13.61 GiB (-28.46%) |
| Peak optimizer state | 11.59 GiB | 3.47 GiB | -8.12 GiB (-70.05%) |
| Peak process RSS | 38.09 GiB | 16.07 GiB | -22.02 GiB |
| Training time | 183.18 s | 80.31 s | -56.16% (2.28x faster) |
| Baseline held-out loss | 1.450091 | 1.450091 | identical |
| Post-training held-out loss | 0.998779 | 0.989205 | -0.009574 |

The parameter-count assertion matches to the unit:

- `true`: 1,089,998,336 always-on parameters + 466,115,584 sampled decoder parameters
  = 1,556,113,920 trainable parameters.
- `false`: 0 always-on parameters + 466,115,584 sampled decoder parameters
  = 466,115,584 trainable parameters.

Both short runs learned, and freezing the always-on group did not produce an obvious
held-out-loss collapse. The slightly lower loss in the frozen arm must not be read as a
quality win: this is one seed and only 40 steps.

For system-state validation, swap stayed exactly at 689 MiB before, throughout, and after
both arms. The unrelated idle oMLX server stayed at 361–654 MiB RSS and 0.0–0.1% CPU, far
below the approximately 5.1 GiB concurrent-model state that invalidated my earlier run.
The 113.50 GiB driver peak therefore reproduced without that prior memory pressure.

Apple Silicon uses unified memory, so the two MPS allocator figures above are not dedicated
CUDA VRAM. Model master parameters were FP32 and Trainer BF16 autocast was enabled on MPS.
Both arms also ran on battery power, so the memory result is clean but the timing result
should be treated as directional until repeated on AC power with randomized arm order and
multiple seeds.

Validation on the measured commit: the dedicated #377 suite passes (14 tests), and Ruff
passes for the benchmark harness. I have not opened another PR because #576 already
contains the implementation; this comment supplies the remaining live memory/loss evidence.

**Comment by Srinivasan8888** [comment]
Thanks @Amix29 — this is exactly the hardware half I couldn't run, and it closes the two deferred acceptance items cleanly.

Your trainable-parameter counts match the implementation's arithmetic to the unit: `true` = 1,556,113,920 and `false` = 466,115,584, i.e. the always-on group (1,089,998,336) is precisely what gets frozen, leaving only the 466,115,584 sampled-decoder params — same identity the `TestIssue271`-style count assertion in #576 pins. Good call flagging the slightly-lower frozen-arm loss as directional rather than a quality win at one seed / 40 steps.

One note tying it back to #576's preflight behavior: the analytical VRAM preflight still classifies LISA as full fine-tuning regardless of `lisa_train_embeddings` (documented limitation — it never under-predicts, but won't credit this saving). Your ~64% peak-MPS / −70% optimizer-state numbers are the kind of measured evidence a future fitted-constant estimator credit would need, though MPS unified memory isn't directly CUDA VRAM. Between the merged implementation and these measurements, every acceptance criterion here looks satisfied.

**Comment by MakazhanAlpamys** [comment]
Closed after the follow-up measurement on merged PR #576. Credit to @Amix29 for the paired M4 Max validation with identical model, data, and layer schedule; exact trainable-parameter arithmetic; peak-memory, optimizer-state, RSS, timing, and held-out-loss measurements; and explicit caveats about unified memory, battery power, and single-seed timing. The implementation and evidence now cover the requested flag behavior, parameter-count invariant, memory impact, quality result, and documentation. Any later discrepancy can be tracked in a new issue.

**Comment by MakazhanAlpamys** [comment]
Closed after the follow-up measurement on merged PR #576. Credit to @Amix29 for the paired M4 Max validation with identical model, data, and layer schedule; exact trainable-parameter arithmetic; peak-memory, optimizer-state, RSS, timing, and held-out-loss measurements; and explicit caveats about unified memory, battery power, and single-seed timing. The implementation and evidence now cover the requested flag behavior, parameter-count invariant, memory impact, quality result, and documentation. Any later discrepancy can be tracked in a new issue.
